File: my-dashboard/src/Types2/campaign.ts
// src/types/campaign.ts

// حالات الحملة المعتمدة في قاعدة البيانات (Enum)
export type CampaignStatus =
  | "draft"
  | "pending"
  | "approved"
  | "rejected"
  | "ongoing"
  | "completed"
  | "cancelled";

// هيكل بيانات الصورة القادمة من الـ API
export interface CampaignPhoto {
  photoId: number;
  campaignId: number;
  progressId?: number;
  photoUrl: string;
  uploadedAt: string;
}

// هيكل بيانات الحملة بالكامل
// export interface Campaign {
//   id: number;
//   proposedBy: number;
//   approvedBy?: number;
//   managedBy?: number;
//   categoryId: number;
//   title: string;
//   description: string;
//   location: string;
//   start_date: string;
//   end_date: string;
//   max_volunteers: number;
//   current_volunteers: number; // يتم حسابها أو جلبها لتعرف كم متطوع سجل
//   status: CampaignStatus;
//   actual_progress: number; // نسبة الإنجاز الفعلي على الأرض
//   photos: CampaignPhoto[];
// }
export interface Campaign {
  campaignId: number;
  title: string;
  description: string;
  location: string;

  startDate: string;
  endDate: string;

  maxVolunteers: number;
  status: CampaignStatus;

  publishedAt: string | null;
  createdAt: string;
  updatedAt: string;

  proposedBy: number;
  approvedBy: number | null;
  managedBy: number | null;

  category: number;

  photos?: {
    id?: number;
    campaignId?: number;
    photoUrl: string;
  }[];
}

// البيانات المطلوبة حصراً عند إرسال فورم إنشاء حملة جديدة
export interface CreateCampaignInput {
  title: string;
  description: string;
  location: string;
  categoryId: number;
  max_volunteers: number;
  start_date: string;
  end_date: string;
}

export interface PaginatedResponse<T> {
  content: T[];
  totalPages: number;
  totalElements: number;
  number: number;
  size: number;
  first: boolean;
  last: boolean;
}

File: src/main/java/com/uni/impact/campaign/Campaign.java
package com.uni.impact.campaign;

import com.uni.impact.application.Application;
import com.uni.impact.attendance.Attendance;
import com.uni.impact.campaign_photo.CampaignPhoto;
import com.uni.impact.category.Category;
import com.uni.impact.progress.Progress;
import com.uni.impact.user.User;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.HashSet;
import java.util.Set;


@Entity
@EntityListeners(AuditingEntityListener.class)
@Getter
@Setter
public class Campaign {

    @Id
    @Column(updatable = false)
    @GeneratedValue(strategy = GenerationType.SEQUENCE)
    private Long campaignId;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "text")
    private String description;

    @Column
    private String location;

    @Column(nullable = false)
    private LocalDate startDate;

    @Column(nullable = false)
    private LocalDate endDate;

    @Column
    private Integer maxVolunteers;

    @Column(nullable = false, length = 20)
    @Enumerated(EnumType.STRING)
    private CampaignStatus status;

    @Column
    private OffsetDateTime publishedAt;

    @Column(nullable = false, updatable = false)
    @CreatedDate
    private OffsetDateTime createdAt;

    @Column(nullable = false)
    @LastModifiedDate
    private OffsetDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "proposed_by_id", nullable = false)
    private User proposedBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "approved_by_id")
    private User approvedBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "managed_by_id")
    private User managedBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @OneToMany(mappedBy = "campaign")
    private Set<CampaignPhoto> campaignCampaignPhotos = new HashSet<>();

    @OneToMany(mappedBy = "campaign")
    private Set<Progress> campaignProgresses = new HashSet<>();

    @OneToMany(mappedBy = "campaign")
    private Set<Application> campaignApplications = new HashSet<>();

    @OneToMany(mappedBy = "campaign")
    private Set<Attendance> campaignAttendances = new HashSet<>();

}
